Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m

```diff
@@ -0,0 +1,356 @@
+% DEVELOPER PM:
+% 
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION: This function performs subtraction and averaging all across the ExploreASL ASL modules, such as xASL_wrp_Realign,
+%              and xASL_wrp_RegisterASL to compute temporary PWI images, xASL_wrp_ResampleASL for creating QC images and preparing
+%              quantification, and xASL_wrp_Quantify if PWI4D is not used by the quantification algorithm (such as BASIL)
+%
+% This function performs the following steps:
+% 1. Admin & checks
+% 2. Apply motion correction if this is needed
+%              
+% Nomenclature:
+% ASL4D = original timeseries
+% PWI4D = subtracted timeseries -> for quantification purposes
+% PWI3D = averaged per TE—PLD—LabDur combination -> for QC purposes
+% PWI   = single perfusion-weighted volume -> for registration purposes
+
+% The quantification submodule xASL_wrp_Quantify only uses PWI4D from here, and recalculates PWI3D and PWI for quantification, potentially improving
+% quantification. So the user can, after QC, potentially remove some volumes from PWI4D, but should be aware that the information in the JSON sidecar
+% (TE, PLD, LabDur) should be adapted accordingly.
+%
+% EXAMPLE used by xASL_wrp_RegisterASL: PWI = xASL_im_ASLSubtractionAveraging(x, ASL_im);
+% EXAMPLE used by xASL_wrp_ResampleASL: [PWI, PWI3D, PWI4D, x] = xASL_im_ASLSubtractionAveraging(x, ASL4D, PWI4D, PWI3D);
+% __________________________________
+% Copyright (C) 2015-2023 ExploreASL
+
+
+%% 1. Admin & checks
+% Input image volumes should be correct
+% Creation of the PWI3D or PWI4D image volumes can be skipped if they are provided as input
+bCreatePWI3D = true;
+bCreatePWI4D = true;
+
+
+if nargin==4
+    if isempty(PWI3D) || ~isnumeric(PWI3D)
+        error('Illegal PWI3D');
+    elseif numel(size(PWI3D))>4 || numel(size(PWI3D))<3
+        error('PWI3D has incorrect number of dimensions');
+    else
+        bCreatePWI3D = false;
+    end
+end
+if nargin==3
+    if isempty(PWI4D) || ~isnumeric(PWI4D)
+        error('Illegal PWI4D');
+    elseif numel(size(PWI4D))~=4
+        error('PWI4D has incorrect number of dimensions');
+    else
+        bCreatePWI4D = false;
+    end
+end
+
+if bCreatePWI4D
+
+    if nargin<2 || isempty(path_ASL4D)
+        error('path_ASL4D input missing');
+    else
+        ASL4D = xASL_io_Nifti2Im(path_ASL4D); % Load time-series nifti from path if input was a path
+        % (input can also be an image matrix)
+        nVolumes = size(ASL4D, 4);
+        
+        
+        % =====================================================================
+        if numel(size(ASL4D))>4
+            warning('In BIDS ASL NIfTIs should have [X Y Z n/PLD/TE/etc] as 4 dimensions');
+            error('ASL4D NIfTI has more than 4 dimensions');
+            
+        elseif nVolumes>1 && mod(nVolumes, 2) ~= 0
```

Comment:
See my previous comment. Odd number of volumes has to be allowed in case that x.modules.asl.bContainsDeltaM is true.

Reply:
Done in 35bafbb72fa5eecffb52fab7fe4079a4db8f6c5c